src/Microsoft.VisualStudio.Services.Agent/ProcessInvoker.cs

```diff
@@ -176,14 +145,43 @@ public void Dispose()
             GC.SuppressFinalize(this);
         }
 
-
         private void Dispose(bool disposing)
         {
             if (disposing)
             {
                 _processExitedSignal.Dispose();
-                _proc?.Dispose();
-                _proc = null;
+                if (_proc != null)
+                {
+                    _proc.Exited -= OnExited;
+                    _proc.ErrorDataReceived -= OnErrorDataReceived;
+                    _proc.OutputDataReceived -= OnOutputDataReceived;
+                    _proc.Dispose();
+                    _proc = null;
+                }
+            }
+        }
+
+        private void OnExited(object sender, EventArgs e)
+        {
+            _stopWatch.Stop();
+            _processExitedSignal.Release();
+        }
+
+        private void OnErrorDataReceived(object sender, DataReceivedEventArgs e)
+        {
+            // at the end of the process, the event fires one last time with null
+            if (e.Data != null)
+            {
+                ErrorDataReceived?.Invoke(sender, e);
```

Comment:
I removed the synchronization lock. I have an idea for a lock-less job server queue. I didn't want an unnecessary sync lock here to slow down handling logging spewing out of a build. Caller should handle. Monitor.Enter isn't so cheap that we should force it at this low level (in process invoker).